experimental/unicodeset_parser/src/parse.rs

```diff
@@ -0,0 +1,1168 @@
+    /// ```
+    pub fn fmt_with_source(&self, source: &str) -> impl Display {
+        let ParseError { offset, kind } = *self;
+
+        if kind == ParseErrorKind::Eof {
+            return format!("{source}<-- error: unexpected end of input");
+        }
+        let mut s = String::new();
+        if let Some(offset) = offset {
+            if offset < source.len() {
+                // offset points to any byte of the last character we want to display.
+                // in the case of ASCII, this is easy - we just display bytes [..=offset].
+                // however, if the last character is more than one byte in UTF-8
+                // we cannot use ..=offset, because that would potentially include only partial
+                // bytes of last character in our string. hence we must find the start of the
+                // following character and use that as the (exclusive) end of our string.
+
+                // offset points into the last character we want to include, hence the start of the
+                // first character we want to exclude is at least offset + 1.
+                let mut exclusive_end = offset + 1;
+                // replace this loop with str::ceil_char_boundary once stable
+                for _ in 0..3 {
+                    // is_char_boundary returns true at the latest once exclusive_end == source.len()
+                    if source.is_char_boundary(exclusive_end) {
+                        break;
+                    }
+                    exclusive_end += 1;
+                }
+
+                // exclusive_end is at most source.len() due to str::is_char_boundary and at least 0 by type
+                #[allow(clippy::indexing_slicing)]
+                s.push_str(&source[..exclusive_end]);
+                s.push_str("<-- ");
+            }
+        }
+        s.push_str("error: ");
+        match kind {
+            ParseErrorKind::UnexpectedChar(c) => {
+                s.push_str(&format!("unexpected character '{}'", c.escape_debug()));
+            }
+            ParseErrorKind::UnknownProperty => {
+                s.push_str("unknown property");
+            }
+            ParseErrorKind::Eof => {
+                s.push_str("unexpected end of input");
+            }
+            ParseErrorKind::Internal => {
+                s.push_str("internal error");
+            }
+            ParseErrorKind::Unimplemented => {
+                s.push_str("unimplemented");
+            }
+            ParseErrorKind::InvalidEscape => {
+                s.push_str("invalid escape sequence");
+            }
+        }
+
+        s
+    }
+
+    /// Returns the [`ParseErrorKind`] of this error.
+    pub fn kind(&self) -> ParseErrorKind {
+        self.kind
+    }
+}
+
+// necessary helper because char::is_whitespace is not equivalent to [:Pattern_White_Space:]
+#[inline]
+fn is_char_pattern_white_space(c: char) -> bool {
```

Comment:
`icu_properties::sets::load_pattern_white_space`, but yes! At the moment the parser unconditionally depends on this data, but I suppose we could turn white-space checking into a UnicodeSetBuilderOption and only if true we load the set in the eventual `UnicodeSetBuilder::try_new_unstable` and store the payload on the builder or return on error?

Reply:
I lean toward hard coding / baking this property, since it is tiny, unlikely to change, and needed for every parse, and having a test to ensure it stays up to date

Reply:
I'll leave it as is for now.